File: codes/lua/Array.includes.lua
-- There's no JavaScript-like Array.includes() in Lua.
-- But, we can create our own function to mimic it in Lua.

function pretty_array_of_primitives(an_array_of_primitives)
    local result = "["
    for array_item_index, array_item in ipairs(an_array_of_primitives) do
        if type(array_item) ~= "string" and type(array_item) ~= "number" then
            goto next
        end
        if type(array_item) == "string" then
            result = result .. "\"" .. array_item .. "\""
        end
        if type(array_item) == "number" then
            result = result .. array_item
        end
        if array_item_index ~= #an_array_of_primitives then
            result = result .. ", "
        end
        ::next::
    end
    result = result .. "]"
    return result
end

function array_includes_v1(search_element, an_array)
    -- JavaScript-like Array.includes() function
    local element_found = false
    for _, array_item in ipairs(an_array) do
        local is_condition_match = array_item == search_element
        if is_condition_match == true then
            element_found = true
            break
        end
    end
    return element_found
end

function array_includes_v2(search_element, an_array)
    -- JavaScript-like Array.includes() function
    local element_found = false
    for _, array_item in ipairs(an_array) do
        if array_item == search_element then
            element_found = true
            break
        end
    end
    return element_found
end

function array_includes_v3(search_element, an_array)
    -- JavaScript-like Array.includes() function
    local element_found = false
    for _, array_item in ipairs(an_array) do
        local is_condition_match = array_item == search_element
        if is_condition_match == true then
            element_found = true
            return element_found
        end
    end
    return element_found
end

function array_includes_v4(search_element, an_array)
    -- JavaScript-like Array.includes() function
    local element_found = false
    for _, array_item in ipairs(an_array) do
        if array_item == search_element then
            element_found = true
            return element_found
        end
    end
    return element_found
end

function array_includes_v5(search_element, an_array)
    -- JavaScript-like Array.includes() function
    for _, array_item in ipairs(an_array) do
        local is_condition_match = array_item == search_element
        if is_condition_match == true then
            return true
        end
    end
    return false
end

function array_includes_v6(search_element, an_array)
    -- JavaScript-like Array.includes() function
    for _, array_item in ipairs(an_array) do
        if array_item == search_element then
            return true
        end
    end
    return false
end

print('\n-- JavaScript-like Array.includes() in Lua')

my_friends = {"Alisa", "Trivia"}
print("my friends: " .. pretty_array_of_primitives(my_friends))

print("-- using JavaScript-like Array.includes() function \"array_includes_v1\"")

name = "Alisa"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v1(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Alisa": true

name = "Trivia"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v1(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Trivia": true

name = "Tony"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v1(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Tony": false

name = "Ezekiel"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v1(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Ezekiel": false

print("-- using JavaScript-like Array.includes() function \"array_includes_v2\"")

name = "Alisa"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v2(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Alisa": true

name = "Trivia"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v2(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Trivia": true

name = "Tony"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v2(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Tony": false

name = "Ezekiel"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v2(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Ezekiel": false

print("-- using JavaScript-like Array.includes() function \"array_includes_v3\"")

name = "Alisa"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v3(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Alisa": true

name = "Trivia"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v3(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Trivia": true

name = "Tony"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v3(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Tony": false

name = "Ezekiel"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v3(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Ezekiel": false

print("-- using JavaScript-like Array.includes() function \"array_includes_v4\"")

name = "Alisa"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v4(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Alisa": true

name = "Trivia"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v4(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Trivia": true

name = "Tony"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v4(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Tony": false

name = "Ezekiel"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v4(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Ezekiel": false

print("-- using JavaScript-like Array.includes() function \"array_includes_v5\"")

name = "Alisa"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v5(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Alisa": true

name = "Trivia"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v5(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Trivia": true

name = "Tony"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v5(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Tony": false

name = "Ezekiel"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v5(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Ezekiel": false

print("-- using JavaScript-like Array.includes() function \"array_includes_v6\"")

name = "Alisa"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v6(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Alisa": true

name = "Trivia"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v6(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Trivia": true

name = "Tony"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v6(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Tony": false

name = "Ezekiel"
print("name: " .. "\"" .. name .. "\"")
is_my_friend = array_includes_v6(name, my_friends)
print("is my friends includes " .. "\"" .. name .. "\"" .. ":" .. tostring(is_my_friend))
-- is my friends includes "Ezekiel": false

File: codes/r/Array.includes.r
library(jsonlite)

# There's no JavaScript-like Array.includes() in R.
# But, we can create our own function to mimic it in R.

prettyJsonStringify <- function(anything) (if (is.null(anything) == TRUE) "NULL" else if (is.character(anything) == TRUE) paste(sep = "", "\"", anything, "\"") else prettify(toJSON(anything, pretty = TRUE, auto_unbox = TRUE), indent = 4))

prettyArrayOfPrimitives <- function(anArrayOfPrimitives) {
    result <- "["
    for (arrayItemIndex in seq_along(anArrayOfPrimitives)) {
        arrayItem <- anArrayOfPrimitives[[arrayItemIndex]]
        if ((is.character(arrayItem) == FALSE) && (is.numeric(arrayItem) == FALSE) && (is.logical(arrayItem) == FALSE) && (is.null(arrayItem) == FALSE)) next
        if (is.character(arrayItem) == TRUE) {
            result <- paste(sep = "", result, "\"", arrayItem, "\"")
        }
        if ((is.numeric(arrayItem) == TRUE) || (is.logical(arrayItem) == TRUE)) {
            result <- paste(sep = "", result, arrayItem)
        }
        if (is.null(arrayItem) == TRUE) {
            result <- paste(sep = "", result, "NULL")
        }
        if (arrayItemIndex != length(anArrayOfPrimitives)) {
            result <- paste(sep = "", result, ", ")
        }
    }
    result <- paste(sep = "", result, "]")
    return(result)
}

arrayIncludesV1 <- function(searchElement, anArray) {
    # JavaScript-like Array.includes() function
    elementFound <- FALSE
    for (arrayItemIndex in seq_along(anArray)) {
        arrayItem <- anArray[[arrayItemIndex]]
        isConditionMatch <- (arrayItem == searchElement)
        if (isConditionMatch == TRUE) {
            elementFound <- TRUE
            break
        }
    }
    return(elementFound)
}

arrayIncludesV2 <- function(searchElement, anArray) {
    # JavaScript-like Array.includes() function
    elementFound <- FALSE
    for (arrayItemIndex in seq_along(anArray)) {
        arrayItem <- anArray[[arrayItemIndex]]
        if (arrayItem == searchElement) {
            elementFound <- TRUE
            break
        }
    }
    return(elementFound)
}

arrayIncludesV3 <- function(searchElement, anArray) {
    # JavaScript-like Array.includes() function
    elementFound <- FALSE
    for (arrayItemIndex in seq_along(anArray)) {
        arrayItem <- anArray[[arrayItemIndex]]
        isConditionMatch <- (arrayItem == searchElement)
        if (isConditionMatch == TRUE) {
            elementFound <- TRUE
            return(elementFound)
        }
    }
    return(elementFound)
}

arrayIncludesV4 <- function(searchElement, anArray) {
    # JavaScript-like Array.includes() function
    elementFound <- FALSE
    for (arrayItemIndex in seq_along(anArray)) {
        arrayItem <- anArray[[arrayItemIndex]]
        if (arrayItem == searchElement) {
            elementFound <- TRUE
            return(elementFound)
        }
    }
    return(elementFound)
}

arrayIncludesV5 <- function(searchElement, anArray) {
    # JavaScript-like Array.includes() function
    for (arrayItemIndex in seq_along(anArray)) {
        arrayItem <- anArray[[arrayItemIndex]]
        isConditionMatch <- (arrayItem == searchElement)
        if (isConditionMatch == TRUE) return(TRUE)
    }
    return(FALSE)
}

arrayIncludesV6 <- function(searchElement, anArray) {
    # JavaScript-like Array.includes() function
    for (arrayItemIndex in seq_along(anArray)) {
        arrayItem <- anArray[[arrayItemIndex]]
        if (arrayItem == searchElement) return(TRUE)
    }
    return(FALSE)
}

cat("\n# JavaScript-like Array.includes() in R\n")

myFriends <- list("Alisa", "Trivia")
cat(paste(sep = "", "my friends: ", prettyArrayOfPrimitives(myFriends), "\n"))

cat("# using JavaScript-like Array.includes() function \"arrayIncludesV1\"\n")

name <- "Alisa"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV1(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Alisa": TRUE

name <- "Trivia"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV1(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Trivia": TRUE

name <- "Tony"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV1(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Tony": FALSE

name <- "Ezekiel"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV1(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Ezekiel": FALSE

cat("# using JavaScript-like Array.includes() function \"arrayIncludesV2\"\n")

name <- "Alisa"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV2(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Alisa": TRUE

name <- "Trivia"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV2(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Trivia": TRUE

name <- "Tony"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV2(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Tony": FALSE

name <- "Ezekiel"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV2(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Ezekiel": FALSE

cat("# using JavaScript-like Array.includes() function \"arrayIncludesV3\"\n")

name <- "Alisa"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV3(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Alisa": TRUE

name <- "Trivia"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV3(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Trivia": TRUE

name <- "Tony"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV3(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Tony": FALSE

name <- "Ezekiel"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV3(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Ezekiel": FALSE

cat("# using JavaScript-like Array.includes() function \"arrayIncludesV4\"\n")

name <- "Alisa"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV4(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Alisa": TRUE

name <- "Trivia"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV4(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Trivia": TRUE

name <- "Tony"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV4(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Tony": FALSE

name <- "Ezekiel"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV4(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Ezekiel": FALSE

cat("# using JavaScript-like Array.includes() function \"arrayIncludesV5\"\n")

name <- "Alisa"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV5(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Alisa": TRUE

name <- "Trivia"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV5(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Trivia": TRUE

name <- "Tony"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV5(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Tony": FALSE

name <- "Ezekiel"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV5(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Ezekiel": FALSE

cat("# using JavaScript-like Array.includes() function \"arrayIncludesV6\"\n")

name <- "Alisa"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV6(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Alisa": TRUE

name <- "Trivia"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV6(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Trivia": TRUE

name <- "Tony"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV6(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Tony": FALSE

name <- "Ezekiel"
cat(paste(sep = "", "name: ", prettyJsonStringify(name), "\n"))
isMyFriend <- arrayIncludesV6(name, myFriends)
cat(paste(sep = "", "is my friends includes ", prettyJsonStringify(name), ": ", isMyFriend, "\n"))
# is my friends includes "Ezekiel": FALSE
